### DISCUSSION TITLE
What if i have to pass a constructor parameter to viewmodel?

### DISCUSSION THREAD

**Discussion by yuvaraj119** [arkivanov/Decompose#771] [Q&A]
> Thank you for the report and the reproducer. It appears to be a bug in the documentation. The official navigation clears the child `ViewModelStore` when the screen is removed from the composition, whereas Decompose clears on `pop`. As a result, a new instance of the child ViewModel is created when the child screen recomposes because of the exit animation.
> 
> Please use the following function for creating ViewModels. I will also need to update the docs.
> 
> ```kotlin
> @Composable
> inline fun <reified T : ViewModel> rememberViewModel(): T {
>     var vm by remember { mutableStateOf<T?>(null) }
> 
>     if (vm == null) {
>         vm = koinViewModel<T>()
>     }
> 
>     return requireNotNull(vm)
> }
> ```

What if i have to pass a constructor parameter to viewmodel?
something like this 
**First approach:**
`class XScreenViewModel(val iXScreen: IXScreen):ViewModel() {
}`

OR 
**Second approach**
If i make my viewmodel component will that cause issue?

`class SplashScreenViewModel(componentContext: ComponentContext) : ViewModel(), IXScreen,
    ComponentContext by componentContext {`
    
best would be First if it works. This will isolate the whole decompose component from actual code and later at somepoint if we have to switch the Navigation it will be easy to do.

**Comment by arkivanov** [comment]
> What if i have to pass a constructor parameter to viewmodel?

You can either pass the dependencies manually or use a DI framework (like `koinViewModel()` in the snippet you provided).

> If i make my viewmodel component will that cause issue?

Usually, ViewModels are not needed with Decompose. If you need to survive configuration changes on Android: see the [doc](https://arkivanov.github.io/Decompose/component/instance-retaining/).
